File: Reqnroll/Chapter15/15.7-TheStubDependencyPattern/After/WIMP.Specs/Support/StubTimeService.cs
using WIMP.App.Services;

namespace WIMP.Specs.Support;

public class StubTimeService : ITimeService
{
    private DateTimeOffset now = DateTimeOffset.Now;
    private readonly List<Func<DateTimeOffset, bool>> timeChangeSubscribers = [];

    public DateTimeOffset GetCurrentTime() => now;

    public void SubscribeToTimeChange(Func<DateTimeOffset, bool> onTimeChanged) =>
        timeChangeSubscribers.Add(onTimeChanged);

    public void SetCurrentTime(DateTimeOffset currentDateTime)
    {
        now = currentDateTime;
        TriggerTimeChange();
    }

    public void TriggerTimeChange()
    {
        var currentDateTime = GetCurrentTime();
        foreach (var subscriber in timeChangeSubscribers.ToArray())
        {
            if (subscriber(currentDateTime))
            {
                timeChangeSubscribers.Remove(subscriber);
            }
        }
    }
}

File: CucumberJVM/Chapter15/15.7-StubDependencyPattern/After/src/test/java/com/wimp/specs/support/StubTimeService.java
package com.wimp.specs.support;

import com.wimp.app.services.TimeService;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class StubTimeService implements TimeService {
    private LocalDateTime now = LocalDateTime.now();
    private final List<Predicate<LocalDateTime>> timeChangeSubscribers = new ArrayList<>();

    @Override
    public LocalDateTime getCurrentTime() {
        return now;
    }

    @Override
    public void subscribeToTimeChange(Predicate<LocalDateTime> onTimeChanged) {
        timeChangeSubscribers.add(onTimeChanged);
    }

    public void setCurrentTime(LocalDateTime currentDateTime) {
        now = currentDateTime;
        triggerTimeChange();
    }

    public void triggerTimeChange() {
        LocalDateTime currentDateTime = getCurrentTime();
        for (Predicate<LocalDateTime> subscriber : List.copyOf(timeChangeSubscribers)) {
            if (subscriber.test(currentDateTime)) {
                timeChangeSubscribers.remove(subscriber);
            }
        }
    }
}
